### PR TITLE
Replace DoubleByteBuf with CompositeByteBuf because of perf regression with Netty > 4.1.12

### PR DESCRIPTION
**Pull request by merlimat**
Starting from Netty-4.1.13 some internal behavior has changed and that has introduced a performance penalty when using the `DoubleByteBuf`. 

The problem resides in the fact that netty falls back to calling `nioBuffers()` and that is creating an array of unpooled direct `ByteBuffers`. That is pretty heavy and the JVM is taking 1 to 3 seconds GC pause to reclaim them. 

Short term fix is to go back to `CompositeByteBuf` which doesn't have the problem (the `DirectByteBuffer` instances are pooled and reused). I have verified that GC is back to normal after this patch.

More mid-term, since `CompositeByteBuf` allocates a bunch of objects (counted ~10 per entry) I have a change to replace it with a `ByteBufList` holder class. I have the change almost done but it will take a while to test it to verify corner cases.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java [hunk 1]
```diff
@@ -17,377 +17,21 @@
...
-
-    @Override
-    public ByteBuffer internalNioBuffer(int index, int length) {
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public ByteBuffer nioBuffer(int index, int length) {
-        ByteBuffer dst = isDirect() ? ByteBuffer.allocateDirect(length) : ByteBuffer.allocate(length);
-        ByteBuf b = Unpooled.wrappedBuffer(dst);
```
Comment (rdhabalia) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java]: so, is this the cause for allocating unpooled direct ByteBuffer?

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java]: Yes, until netty-4.1.12 it was going through the `DoubleByteBuf.getBytes()` and so there were no allocations.

**Comment by rdhabalia** [comment]
@saandrews I have created PR in internal branch: can you approve it and then I will promote the change here.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java (hunk 1 above)
Comment (saandrews) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java]: Do we have the option to use older netty in bookie? In that case, we need to keep old code as well.

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java]: @saandrews @rdhabalia  It's not possible to use get one implementation or the other depending on the Netty version at runtime. 

I'm still testing on a fix to get rid of DoubleByteBuf/CompositeByteBuf, but that's not a trivial change, I cannot get it working reliably yet.

I haven't seen any performance degradation going from Pulsar-1.21 to the current master with this patch applied. If you guys are worried, you can still depend on the 4.3.1.85-yahoo BK artifact for the internal deployment, until we merge the final fix.

Right now, the current implementation degrades badly with newer Netty, so we need a quick fix for it.

Comment (saandrews) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java]: I see. Lets proceed with this for now then.

**Comment by rdhabalia** [comment]
@saandrews can you also approve it in internal branch so, we can merge it and promote here.

**Comment by rdhabalia** [comment]
closing it as merged into internal branch.
